File: scripts/fetch-league-averages.mjs
#!/usr/bin/env node
/**
 * Scrapes basketball-reference's team-totals table for each season and
 * derives the league-wide rates the VA formula uses, merging into
 * app/data/league-averages.json. One-shot tool — run when you need to
 * extend coverage backward so the playoff bake can produce meaningful
 * VA for older seasons.
 *
 *   node scripts/fetch-league-averages.mjs <startSeason> [endSeason]
 *
 *   node scripts/fetch-league-averages.mjs 1979-80
 *   node scripts/fetch-league-averages.mjs 1970-71 1995-96
 *
 * Pulls from:
 *   https://www.basketball-reference.com/leagues/NBA_<endYear>.html
 *
 * Existing entries are preserved unless --force is passed.
 */

import { writeFile, readFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import * as cheerio from "cheerio";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const DATA_DIR = join(ROOT, "app", "data");
const LGA_PATH = join(DATA_DIR, "league-averages.json");

const UA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36";
const REQUEST_DELAY_MS = 2500;
let lastRequest = 0;

const SEASON_RE = /^\d{4}-\d{2}$/;

const args = process.argv.slice(2);
const force = args.includes("--force");
const positional = args.filter((a) => !a.startsWith("--"));
if (positional.length < 1 || positional.length > 2) {
  console.error("Usage: node scripts/fetch-league-averages.mjs <startSeason> [endSeason] [--force]");
  process.exit(1);
}
const [startSeason, endSeason = positional[0]] = positional;
for (const s of [startSeason, endSeason]) {
  if (!SEASON_RE.test(s)) {
    console.error(`Bad season "${s}" — expected YYYY-YY`);
    process.exit(1);
  }
}
const startYear = Number(startSeason.slice(0, 4));
const endYear = Number(endSeason.slice(0, 4));
if (endYear < startYear) {
  console.error(`endSeason (${endSeason}) is before startSeason (${startSeason})`);
  process.exit(1);
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const num = (v) => {
  const n = parseFloat(String(v).trim());
  return Number.isFinite(n) ? n : 0;
};

async function throttledFetch(url) {
  const wait = Math.max(0, REQUEST_DELAY_MS - (Date.now() - lastRequest));
  if (wait > 0) await sleep(wait);
  lastRequest = Date.now();
  const res = await fetch(url, { headers: { "User-Agent": UA, Accept: "text/html" } });
  if (res.status === 429 || res.status === 503) {
    console.warn(`  ${res.status} on ${url} — sleeping 60s`);
    await sleep(60_000);
    return throttledFetch(url);
  }
  if (!res.ok) throw new Error(`HTTP ${res.status} ${url}`);
  return res.text();
}

// Find the team-totals table on the season page. BR's table ids have
// churned (totals-team, team_totals, team-stats-base). Fall back to any
// table whose <thead> shows fg/fga and mp columns and whose tbody has
// roughly the right number of rows (between 8 and 60 — covers ABA-era
// merged tables and modern 30-team seasons).
function findTotalsTable($) {
  const ids = ["totals-team", "team_totals", "team-stats-base", "totals_team"];
  for (const id of ids) {
    const t = $(`table#${id}`);
    if (t.length) return t;
  }
  let found = null;
  $("table").each((_, el) => {
    if (found) return;
    const t = $(el);
    const head = t.find("thead");
    if (!head.length) return;
    const hasFga = head.find("[data-stat='fga']").length > 0;
    const hasMp = head.find("[data-stat='mp']").length > 0;
    const hasPts = head.find("[data-stat='pts']").length > 0;
    const rowCount = t.find("tbody tr").length;
    if (hasFga && hasMp && hasPts && rowCount >= 8 && rowCount <= 60) found = t;
  });
  return found;
}

async function fetchLeagueTotals(season) {
  const yearEnd = Number(season.slice(0, 4)) + 1;
  const url = `https://www.basketball-reference.com/leagues/NBA_${yearEnd}.html`;
  console.log(`Fetching ${url}`);
  const html = await throttledFetch(url);
  const $ = cheerio.load(html.replace(/<!--([\s\S]*?)-->/g, "$1"));
  const table = findTotalsTable($);
  if (!table || !table.length) throw new Error("team totals table not found");

  const totals = {
    mp: 0, pts: 0, ast: 0, stl: 0, blk: 0, tov: 0, drb: 0, orb: 0,
    fgm: 0, fga: 0, tpm: 0, tpa: 0, ftm: 0, fta: 0,
  };
  let rows = 0;
  table.find("tbody tr").each((_, tr) => {
    const $tr = $(tr);
    if ($tr.hasClass("thead")) return;
    const cell = (...keys) => {
      for (const k of keys) {
        const v = $tr.find(`[data-stat='${k}']`).first().text().trim();
        if (v !== "") return v;
      }
      return "";
    };
    const teamName = cell("team", "team_id", "team_name", "team_name_abbr");
    // Skip "League Average" / "League Total" summary rows; we sum the
    // per-team rows ourselves to keep this consistent.
    if (/league/i.test(teamName)) return;
    const mp = num(cell("mp"));
    if (mp <= 0) return;
    rows++;
    totals.mp  += mp;
    totals.pts += num(cell("pts"));
    totals.ast += num(cell("ast"));
    totals.stl += num(cell("stl"));
    totals.blk += num(cell("blk"));
    totals.tov += num(cell("tov"));
    totals.drb += num(cell("drb"));
    totals.orb += num(cell("orb"));
    totals.fgm += num(cell("fg"));
    totals.fga += num(cell("fga"));
    totals.tpm += num(cell("fg3"));
    totals.tpa += num(cell("fg3a"));
    totals.ftm += num(cell("ft"));
    totals.fta += num(cell("fta"));
  });
  if (rows < 8) throw new Error(`only ${rows} team rows parsed; table layout may have changed`);
  return totals;
}

function lgaFromTotals(t) {
  const safe = (a, b) => (b > 0 ? a / b : 0);
  const twoPm = t.fgm - t.tpm;
  const twoPa = t.fga - t.tpa;
  const reb = t.drb + t.orb;
  // Standard possessions estimate (Hollinger): FGA - ORB + TO + 0.475*FTA.
  const poss = t.fga - t.orb + t.tov + 0.475 * t.fta;
  return {
    la3P: safe(t.tpm, t.tpa),
    la2P: safe(twoPm, twoPa),
    laFT: safe(t.ftm, t.fta),
    laFG: safe(t.fgm, t.fga),
    laPTSperM: safe(t.pts, t.mp),
    laASTperM: safe(t.ast, t.mp),
    laSTLperM: safe(t.stl, t.mp),
    laBLKperM: safe(t.blk, t.mp),
    laTOVperM: safe(t.tov, t.mp),
    laDRBperM: safe(t.drb, t.mp),
    laORBperM: safe(t.orb, t.mp),
    laPTSperMake: safe(t.pts, t.fgm),
    laPTSperPoss: safe(t.pts, poss),
    laDRBrate: safe(t.drb, reb),
    laORBrate: safe(t.orb, reb),
  };
}

async function main() {
  let existing = {};
  try {
    existing = JSON.parse(await readFile(LGA_PATH, "utf8"));
  } catch {
    console.log("No existing league-averages.json — starting fresh");
  }

  const seasons = [];
  for (let y = startYear; y <= endYear; y++) {
    const end = String((y + 1) % 100).padStart(2, "0");
    seasons.push(`${y}-${end}`);
  }

  let added = 0, skipped = 0, failed = 0;
  for (const season of seasons) {
    if (existing[season] && !force) {
      console.log(`  Skipping ${season} (already present; pass --force to overwrite)`);
      skipped++;
      continue;
    }
    try {
      const totals = await fetchLeagueTotals(season);
      const lga = lgaFromTotals(totals);
      existing[season] = lga;
      added++;
      console.log(`  ✓ ${season} — laPTSperM=${lga.laPTSperM.toFixed(3)}, la3P=${lga.la3P.toFixed(3)}, laFG=${lga.laFG.toFixed(3)}`);
    } catch (e) {
      console.warn(`  ✗ ${season} — ${e.message}`);
      failed++;
    }
  }

  // Sort keys for a clean diff.
  const sorted = Object.fromEntries(Object.keys(existing).sort().map((k) => [k, existing[k]]));
  await writeFile(LGA_PATH, JSON.stringify(sorted, null, 2) + "\n");
  console.log(`Wrote ${LGA_PATH} (${Object.keys(sorted).length} seasons; +${added} new, ${skipped} skipped, ${failed} failed)`);
}

main().catch((e) => {
  console.error("\n========== FETCH FAILED ==========");
  console.error(e?.stack || e?.message || String(e));
  console.error("==================================\n");
  process.exit(1);
});

File: scripts/R/fetch_league_averages.R
#!/usr/bin/env Rscript
# Fetches basketball-reference's PLAYER season-totals page for each season and
# derives the league-wide rates the VA formula needs, merging into
# app/data/league-averages.json. Existing entries are preserved unless
# --force is passed.
#
#   Rscript scripts/R/fetch_league_averages.R 1979-80
#   Rscript scripts/R/fetch_league_averages.R 1970-71 1995-96 --force
#
# Why the player-totals page (NBA_<year>_totals.html) and not the season
# index's team-totals table: (1) the per-minute baselines are minutes-weighted
# MEDIAN player rates (see scrape_common.R::lga_from_players), which need
# per-player rows, not team sums; (2) the team table's id/layout churned
# across eras and the old fallback quietly matched the wrong table, leaving
# 1996-97+ baselines on a different definition than earlier seasons. This is
# the exact page the regular-season bake already parses correctly, and a
# plausibility gate refuses to write junk if the layout ever shifts again.

source(file.path(dirname(sub("^--file=", "",
  grep("^--file=", commandArgs(FALSE), value = TRUE)[1])), "scrape_common.R"))

args <- commandArgs(trailingOnly = TRUE)
force <- "--force" %in% args
positional <- args[!grepl("^--", args)]
if (length(positional) < 1 || length(positional) > 2) {
  stop("Usage: Rscript fetch_league_averages.R <startSeason> [endSeason] [--force]")
}
start_season <- positional[1]
end_season   <- if (length(positional) >= 2) positional[2] else positional[1]
SEASON_RE <- "^[0-9]{4}-[0-9]{2}$"
for (s in c(start_season, end_season)) {
  if (!grepl(SEASON_RE, s)) stop(sprintf('Bad season "%s" - expected YYYY-YY', s))
}
start_year <- as.integer(substr(start_season, 1, 4))
end_year   <- as.integer(substr(end_season, 1, 4))
if (end_year < start_year) {
  stop(sprintf("endSeason (%s) is before startSeason (%s)", end_season, start_season))
}

# Locate the per-player season-totals table (same id list the regular-season
# bake in fetch_historical.R uses).
find_totals_table <- function(doc) {
  for (id in c("totals_stats", "players_totals", "totals", "per_game_stats")) {
    t <- xml2::xml_find_first(doc, sprintf("//table[@id='%s']", id))
    if (!inherits(t, "xml_missing")) return(t)
  }
  for (t in xml2::xml_find_all(doc, "//table")) {
    head <- xml2::xml_find_first(t, ".//thead")
    if (inherits(head, "xml_missing")) next
    hp <- !inherits(xml2::xml_find_first(head, ".//*[@data-stat='pts']"), "xml_missing")
    hg <- !inherits(xml2::xml_find_first(head, ".//*[@data-stat='g']"),   "xml_missing")
    hm <- !inherits(xml2::xml_find_first(head, ".//*[@data-stat='mp']"),  "xml_missing")
    if (hp && hg && hm) return(t)
  }
  NULL
}

# Every player's season row, one per player: traded players appear as
# per-team rows plus a TOT aggregate, so keep the first row per player (BR
# lists TOT first). Per-player rows (not just sums) because the per-minute
# baselines are minutes-weighted MEDIANS.
fetch_league_players <- function(season) {
  year_end <- season_end_year(season)
  url <- sprintf("https://www.basketball-reference.com/leagues/NBA_%d_totals.html", year_end)
  message(sprintf("Fetching %s", url))
  doc <- parse_html_uncommented(throttled_fetch(url))
  table <- find_totals_table(doc)
  if (is.null(table)) stop("player totals table not found")

  seen <- new.env(parent = emptyenv())
  players <- list()
  for (tr in xml2::xml_find_all(table, ".//tbody/tr")) {
    cls <- xml2::xml_attr(tr, "class")
    if (!is.na(cls) && grepl("thead", cls)) next
    name <- cell_text(tr, c("player", "name_display", "name"))
    if (!nzchar(name) || grepl("league average", name, ignore.case = TRUE)) next
    href <- xml2::xml_attr(xml2::xml_find_first(tr, ".//a[contains(@href,'/players/')]"), "href")
    key <- if (!is.na(href)) href else name
    if (!is.null(seen[[key]])) next
    seen[[key]] <- TRUE
    mp <- num(cell_text(tr, c("mp", "mp_total")))
    if (mp <= 0) next
    players[[length(players) + 1]] <- list(
      mp = mp,
      pts = num(cell_text(tr, "pts")), ast = num(cell_text(tr, "ast")),
      stl = num(cell_text(tr, "stl")), blk = num(cell_text(tr, "blk")),
      tov = num(cell_text(tr, "tov")), drb = num(cell_text(tr, "drb")),
      orb = num(cell_text(tr, "orb")), fgm = num(cell_text(tr, "fg")),
      fga = num(cell_text(tr, "fga")), tpm = num(cell_text(tr, "fg3")),
      tpa = num(cell_text(tr, "fg3a")), ftm = num(cell_text(tr, "ft")),
      fta = num(cell_text(tr, "fta"))
    )
  }
  if (length(players) < 100) {
    stop(sprintf("only %d player rows parsed; table layout may have changed", length(players)))
  }
  players
}

main <- function() {
  existing <- load_league_averages()
  seasons <- vapply(start_year:end_year, make_season, character(1))

  added <- 0; skipped <- 0; failed <- 0
  for (season in seasons) {
    if (!is.null(existing[[season]]) && !force) {
      message(sprintf("  Skipping %s (already present; pass --force to overwrite)", season))
      skipped <- skipped + 1
      next
    }
    # tryCatch RETURNS the rates (NULL on failure); assignment happens out
    # here in main's own scope. (`<<-` inside the expression searches from
    # the global env and misses main's locals — "object not found".)
    res <- tryCatch({
      players <- fetch_league_players(season)
      lga <- lga_from_players(players)
      if (!lga_plausible(lga)) {
        stop(sprintf("implausible laPTSperM=%.4f (expected 0.36-0.56); refusing to write",
                     lga$laPTSperM))
      }
      message(sprintf("  ok %s - laPTSperM=%.3f, la3P=%.3f, laFG=%.3f",
                      season, lga$laPTSperM, lga$la3P, lga$laFG))
      lga
    }, error = function(e) {
      message(sprintf("  x %s - %s", season, conditionMessage(e)))
      NULL
    })
    if (!is.null(res)) {
      existing[[season]] <- res
      added <- added + 1
    } else {
      failed <- failed + 1
    }
  }

  # Sort keys for a clean diff.
  existing <- existing[order(names(existing))]
  write_json_pretty(existing, LGA_PATH)
  message(sprintf("Wrote %s (%d seasons; +%d new, %d skipped, %d failed)",
                  LGA_PATH, length(existing), added, skipped, failed))
}

main()
